pkg/controller/cloneset/scale/cloneset_scale.go

```diff
@@ -0,0 +1,207 @@
+package scale
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	appsv1alpha1 "github.com/openkruise/kruise/pkg/apis/apps/v1alpha1"
+	clonesetutils "github.com/openkruise/kruise/pkg/controller/cloneset/utils"
+	"github.com/openkruise/kruise/pkg/util"
+	"github.com/openkruise/kruise/pkg/util/expectations"
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	kubecontroller "k8s.io/kubernetes/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// LengthOfInstanceID is the length of instance-id
+	LengthOfInstanceID = 5
+
+	// When batching pod creates, initialBatchSize is the size of the initial batch.
+	initialBatchSize = 1
+)
+
+// Interface for managing replicas including create and delete pod/pvc.
+type Interface interface {
+	ManageReplicas(
+		currentCS, updateCS *appsv1alpha1.CloneSet,
+		currentRevision, updateRevision string,
+		pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim,
+	) (bool, error)
+}
+
+// NewScaleControl returns a scale control.
+func NewScaleControl(c client.Client, recorder record.EventRecorder, exp expectations.ScaleExpectations) Interface {
+	return &realControl{Client: c, recorder: recorder, exp: exp}
+}
+
+type realControl struct {
+	client.Client
+	recorder record.EventRecorder
+	exp      expectations.ScaleExpectations
+}
+
+func (r *realControl) ManageReplicas(
+	currentCS, updateCS *appsv1alpha1.CloneSet,
+	currentRevision, updateRevision string,
+	pods []*v1.Pod, pvcs []*v1.PersistentVolumeClaim,
+) (bool, error) {
+	if updateCS.Spec.Replicas == nil {
+		return false, fmt.Errorf("spec.Replicas is nil")
+	}
+
+	if podsToDelete := getPodsToDelete(updateCS, pods); len(podsToDelete) > 0 {
+		return true, r.deletePods(updateCS, podsToDelete, pvcs)
+	}
+
+	var err error
+	diff := len(pods) - int(*updateCS.Spec.Replicas)
+
+	if diff < 0 {
+		// total number of this creation
+		expectedCreations := diff * -1
+		// lack number of current version
+		expectedCurrentCreations := 0
+		if updateCS.Spec.UpdateStrategy.Partition != nil {
+			notUpdatedNum := len(pods) - len(clonesetutils.GetUpdatedPods(pods, updateRevision))
+			expectedCurrentCreations = int(*updateCS.Spec.UpdateStrategy.Partition) - notUpdatedNum
+		}
+
+		// available instance-id come from free pvc
+		availableIDs := getOrGenAvailableIDs(expectedCreations, pods, pvcs)
```

Comment:
I think there is no need to persist pvc  during upgrade that is not in place, at least not this time.